#ifndef WIDGET_H
#define WIDGET_H

#include <QWidget>
#include <QSerialPort>
#include <QSerialPortInfo>
#include <QString>
#include <QTime>
#include <QDebug>

QT_BEGIN_NAMESPACE
namespace Ui { class Widget; }
QT_END_NAMESPACE

#define POINTSNUM 10
#define POINTSNUM1 124
#define POINTSNUM2 60
#define LowerFactor 0.8
#define UpperFactor 1.2

#define minScaleX 0
#define scaleRangeX 13
#define maxScaleX 13





class Widget : public QWidget
{
    Q_OBJECT

public:
    Widget(QWidget *parent = nullptr);
    ~Widget();
    double minScaleY=0.0;
    double scaleRangeY=5.0;
    double maxScaleY=5.0;
    void Paint();
    void Painttest();
    void drawpoint(double data[],int ff);
    void Sleep(int msec);
    QSerialPortInfo *serialinfo;
    QSerialPort *serialport;
    double kx,ky1;
    int pointx,pointy;
    QByteArray buffer;
    QByteArray HeadByte;
    int flag_state = 0;

private slots:
    void on_pushButtonP1_clicked();

    void on_pushButtonP2_clicked();

    void on_pushButtonClear_clicked();

    void on_pushButtonOPEN_clicked();

    void readyread();


    void on_pushButtonOK_clicked();

private:
    Ui::Widget *ui;
    uint64_t sendbyte;
    uint64_t revbyte;
    int flag = 1;
    int firstin = 1;
    double measure_data1[130] = {0.0020,0.0076,0.0106,0.0126,0.0145,0.0162,0.0169,0.0171,0.0172,0.0174,0.0175,0.0176,0.0177,0.0178,0.0180,
                                 0.0180,0.0180,0.0180,0.0180,0.0181,0.0181,0.0181,0.0182,0.0181,0.0181,0.0181,0.0181,0.0182,0.0182,0.0182,
                                 0.0183,0.0183,0.0184,0.0183,0.0183,0.0183,0.0183,0.0184,0.0183,0.0184,0.0184,0.0184,0.0184,0.0184,0.0183,
                                 0.0184,0.0187,0.0184,0.0185,0.0184,0.0184,0.0184,0.0187,0.0186,0.0186,0.0187,0.0187,0.0187,0.0187,0.0187,
                                 0.0187,0.0188,0.0188,0.0188,0.0189,0.0188,0.0188,0.0188,0.0189,0.0189,0.0188,0.0189,0.0188,0.0190,0.0189,
                                 0.0189,0.0190,0.0189,0.0190,0.0190,0.0190,0.0190,0.0191,0.0191,0.0190,0.0191,0.0192,0.0190,0.0191,0.0191,
                                 0.0192,0.0192,0.0192,0.0191,0.0192,0.0192,0.0192,0.0192,0.0192,0.0192,0.0192,0.0192,0.0193,0.0194,0.0194,
                                 0.0195,0.0195,0.0195,0.0195,0.0196,0.0196,0.0196,0.0196,0.0196,0.0197,0.0197,0.0198,0.0198,0.0198,0.0198,
                                 0.0199,0.0200,0.0200,0.0199};
    double measure_data2[130] = {0.0021,0.0170,0.0262,0.0352,0.0442,0.0532,0.0621,0.0709,0.0792,0.0879,0.0962,0.1044,0.1126,0.1205,0.1282,
                                 0.1359,0.1432,0.1505,0.1576,0.1641,0.1705,0.1768,0.1824,0.1880,0.1929,0.1980,0.2023,0.2058,0.2097,0.2120,
                                 0.2151,0.2173,0.2194,0.2212,0.2227,0.2245,0.2255,0.2267,0.2279,0.2291,0.2298,0.2307,0.2316,0.2324,0.2333,
                                 0.2341,0.2352,0.2359,0.2366,0.2373,0.2381,0.2390,0.2399,0.2408,0.2416,0.2424,0.2432,0.2442,0.2453,0.2461,
                                 0.2468,0.2478,0.2487,0.2495,0.2508,0.2510,0.2524,0.2532,0.2541,0.2553,0.2561,0.2569,0.2581,0.2592,0.2602,
                                 0.2611,0.2621,0.2630,0.2642,0.2652,0.2664,0.2675,0.2686,0.2694,0.2708,0.2720,0.2727,0.2742,0.2753,0.2764,
                                 0.2777,0.2783,0.2801,0.2810,0.2823,0.2837,0.2845,0.2857,0.2872,0.2770,0.2813,0.2846,0.2875,0.2902,0.2926,
                                 0.2950,0.2973,0.2997,0.3022,0.3045,0.3066,0.3091,0.3114,0.3137,0.3163,0.3186,0.3209,0.3234,0.3258,0.3284,
                                 0.3308,0.3334,0.3360,0.3386};
    double measure_data3[130] = {0.0024,0.0184,0.0285,0.0385,0.0485,0.0585,0.0684,0.0782,0.0880,0.0977,0.1076,0.1173,0.1267,0.1362,0.1457,
                                 0.1550,0.1644,0.1733,0.1823,0.1912,0.2001,0.2086,0.2174,0.2257,0.2340,0.2423,0.2502,0.2580,0.2662,0.2735,
                                 0.2809,0.2885,0.2958,0.3029,0.3100,0.3168,0.3233,0.3302,0.3364,0.3425,0.3490,0.3548,0.3606,0.3665,0.3719,
                                 0.3772,0.3830,0.3874,0.3923,0.3984,0.4034,0.4083,0.4128,0.4173,0.4216,0.4256,0.4289,0.4327,0.4370,0.4414,
                                 0.4439,0.4474,0.4515,0.4538,0.4583,0.4602,0.4632,0.4668,0.4688,0.4717,0.4754,0.4776,0.4795,0.4818,0.4838,
                                 0.4862,0.4883,0.4905,0.4924,0.4949,0.4972,0.4986,0.4999,0.5002,0.5062,0.5093,0.5098,0.5109,0.5120,0.5131,
                                 0.5143,0.5153,0.5161,0.5174,0.5181,0.5190,0.5250,0.5252,0.5263,0.5592,0.5530,0.5510,0.5500,0.5504,0.5509,
                                 0.5516,0.5527,0.5536,0.5550,0.5580,0.5646,0.5640,0.5634,0.5635,0.5639,0.5641,0.5643,0.5648,0.5649,0.5653,
                                 0.5653,0.5654,0.5750,0.5740};
    double testdata[130] = {0.0024,0.0184,0.0285,0.0385,0.0485,0.0585,0.0684,0.0782,0.0880,0.0977,0.1076,0.1173,0.1267,0.1362,0.1457,0.1550,0.1644,0.1733,0.1823,
                            0.1912,0.2001,0.2086,0.2174,0.2257,0.2340,0.2423,0.2502,0.2580,0.2662,0.2735,0.2809,0.2885,0.2958,0.3029,0.3100,0.3168,0.3233,0.3302,
                            0.3364,0.3425,0.3490,0.3548,0.3606,0.3665,0.3719,0.3772,0.3830,0.3874,0.3923,0.3984,0.4034,0.4083,0.4128,0.4173,0.4216,0.4256,0.4289,
                            0.4327,0.4370,0.4414,0.4439,0.4474,0.4515,0.4538,0.4583,0.4602,0.4632,0.4668,0.4688,0.4717,0.4754,0.4776,0.4795,0.4818,0.4838,0.4862,
                            0.4883,0.4905,0.4924,0.4949,0.4972,0.4986,0.4999,0.5002,0.5062,0.5093,0.5098,0.5109,0.5120,0.5131,0.5143,0.5153,0.5161,0.5174,0.5181,
                            0.5190,0.5250,0.5252,0.5263,0.5592,0.5530,0.5510,0.5500,0.5504,0.5509,0.5516,0.5527,0.5536,0.5550,0.5580,0.5646,0.5640,0.5634,0.5635,
                            0.5639,0.5641,0.5643,0.5648,0.5649,0.5653,0.5653,0.5654,0.5750,0.5740,0,0,0,0,0,0};
    double mea[60] = {-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0025,-0.0023,-0.0025,
                      -0.0025,-0.0024,-0.0026,-0.0025,-0.0025,-0.0025,-0.0025,-0.0024,-0.0022,-0.0013,0.0016,0.0086,0.0234,0.0496,0.0914,0.1500,0.2257,
                      0.3179,0.4223,0.5310,0.5771,0.5936,0.6127,0.6398,0.6564,0.6731,0.7026,0.7366,0.7538,0.7910,0.8245,0.8627,0.8914,0.9258,0.9560,
                      0.9704,1.0483,1.4561,2.0740,2.5921,3.1203,4.0485,4.8669,5.6853};
    double xData[124] = {0,0.084,0.160,0.236,0.313,0.388,0.464,0.540,0.616,0.692,0.768,0.844,0.921,0.997,1.073,1.150,1.227,1.302,1.378,1.455,1.531,1.607,
                         1.683,1.759,1.835,1.912,1.998,2.063,2.140,2.216,2.292,2.368,2.445,2.521,2.597,2.673,2.749,2.826,2.902,2.978,3.054,3.131,3.208,3.360,
                         3.436,3.512,3.588,3.664,3.741,3.824,3.901,3.977,4.053,4.130,4.206,4.282,4.358,4.435,4.511,4.588,4.664,4.740,4.816,4.891,4.967,5.044,
                         5.120,5.196,5.272,5.384,5.425,5.501,5.577,5.653,5.729,5.805,5.882,5.958,6.034,6.107,6.182,6.254,6.337,6.412,6.485,6.564,6.633,6.712,
                         6.790,6.862,6.944,7.025,7.094,7.171,7.245,7.322,7.404,7.478,7.552,7.650,7.802,7.954,8.110,8.254,8.401,8.557,8.704,8.853,9.000,9.154,
                         9.307,9.451,9.600,9.754,9.899,10.052,10.201,10.354,10.501,10.652,10.800,10.954,11.102,11.244};
    double x2Data[60] = {0.000,0.150,0.300,0.450,0.600,0.750,0.900,1.050,1.200,1.350,1.500,1.650,1.800,1.950,2.100,2.250,2.400,2.550,2.700,2.850,3.000,3.150,
                         3.300,3.450,3.600,3.750,3.900,4.050,4.200,4.350,4.500,4.650,4.800,4.950,5.100,5.250,5.400,5.550,5.700,5.850,6.000,6.150,6.300,6.450,
                         6.600,6.750,6.900,7.050,7.200,7.350,7.500,7.650,7.800,7.950,8.100,8.250,8.400,8.550,8.700,8.850
                         };

    unsigned int DigitalTube[16] = {0xC0,0xF9,0xA4,0xB0,0x99,0x92,0x82,0xF8,0x80,0x90,0x88,0x83,0xC6,0xA1,0x86,0x8E};

    int count = 0;
    int width;
    int height;
    QByteArray PasteData;
    QString ReadData;

protected:
    void paintEvent(QPaintEvent *);
    QImage m_image;
    QPainter *m_painter;

};
#endif // WIDGET_H
